### PR TITLE
drivers: ethernet: w5500: MAC change support

### PR DESCRIPTION
**Pull request by rpaat**
[W5500 documentation](https://docs.zephyrproject.org/3.0.0/reference/devicetree/bindings/ethernet/wiznet%2Cw5500.html) mentions local-mac-address and zephyr,random-mac-address support, but in reality couldn't find the implementation. This PR adds this functionality.

Updated w5500_set_config() to support ETHERNET_CONFIG_TYPE_MAC_ADDRESS and change the MAC address at runtime as discussed here #39252. Example usage:
  struct net_if *iface;
  struct ethernet_req_params params;
...
  net_if_down(iface);
  ret = net_mgmt(NET_REQUEST_ETHERNET_SET_MAC_ADDRESS, iface,
              &params, sizeof(struct ethernet_req_params));
  net_if_up(iface);

### CODE REVIEW

File: drivers/ethernet/eth_w5500.c [hunk 1]
```diff
@@ -342,28 +342,54 @@ static int w5500_set_config(const struct device *dev,
...
+			ctx->mac_addr[2], ctx->mac_addr[3],
+			ctx->mac_addr[4], ctx->mac_addr[5]);
+
+		/* Register Ethernet MAC Address with the upper layer */
+		net_if_set_link_addr(ctx->iface, ctx->mac_addr,
+			sizeof(ctx->mac_addr),
+			NET_LINK_ETHERNET);
+		return 0;
+	}
+	default:
```
Comment (parthitce) [inline drivers/ethernet/eth_w5500.c:370]: Instead should this be `ETHERNET_CONFIG_TYPE_PROMISC_MODE` ?

File: drivers/ethernet/eth_w5500.c [hunk 2]
```diff
@@ -507,9 +535,9 @@ static int w5500_init(const struct device *dev)
 			LOG_ERR("Unable to configure GPIO pin %u", config->reset.pin);
 			return -EINVAL;
 		}
-		gpio_pin_set_dt(&config->reset, 0);
-		k_usleep(500);
 		gpio_pin_set_dt(&config->reset, 1);
+		k_usleep(500);
+		gpio_pin_set_dt(&config->reset, 0);
```
Comment (parthitce) [inline drivers/ethernet/eth_w5500.c:540]: please drop this changes as it's already under open discussion.

File: drivers/ethernet/eth_w5500.c [hunk 3]
```diff
@@ -342,28 +342,54 @@ static int w5500_set_config(const struct device *dev,
 {
 	uint8_t mode;
 	uint8_t mr = W5500_S0_MR_MF;
+	struct w5500_runtime *ctx = dev->data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
+	{
```
Comment (parthitce) [inline drivers/ethernet/eth_w5500.c:349]: nit, doesn't needed guarded IMO

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: removed the extra guarding

File: drivers/ethernet/eth_w5500.c (hunk 1 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:368]: Did you mean to add `ETHERNET_CONFIG_TYPE_PROMISC_MODE` as one case option in switch statement, instead of being handled in default block?

I didn't want to touch the existing code as it wasn't fully clear to me why it was implemented in that way in first place. Otherwise I would like to refactor the existing code too.

File: drivers/ethernet/eth_w5500.c (hunk 2 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: Removed this change from my PR and waiting for the outcome in separate #46469

**Comment by parthitce** [comment]
@rpaat new changes must go into the same commit. use git commit --amend and rebase -i

**Comment by rpaat** [comment]
> @rpaat new changes must go into the same commit. use git commit --amend and rebase -i

Thank you for the hint. I hope that I managed to do the squashing properly.

File: drivers/ethernet/eth_w5500.c (hunk 1 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:368]: Can't see my comment here. Will try to write again.

I checked the code and promiscuous mode (MAC filter off) is activated by default in w5500_hw_start().

The bad thing is that whenever w5500_set_config() is called and the type is not  ETHERNET_CONFIG_TYPE_PROMISC_MODE then promiscuous mode is disabled (MAC filter is turned on).

This needs more investigation and I didn't want to change this code with this PR.

File: drivers/ethernet/eth_w5500.c [hunk 4]
```diff
@@ -342,28 +342,52 @@ static int w5500_set_config(const struct device *dev,
 {
 	uint8_t mode;
 	uint8_t mr = W5500_S0_MR_MF;
+	struct w5500_runtime *ctx = dev->data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
```
Comment (lucasgcbkhomp) [inline drivers/ethernet/eth_w5500.c:348]: I've been looking at this and I'm wondering why place this switch here instead of implementing it on the `generate_mac` callback on the `w5500_runtime` struct. It seems to mix the mode with the MAC attribution.

**Comment by carlescufi** [comment]
@rpaat sorry for the huge delay. We will try to get this reviewed and merged.

**Comment by rpaat** [comment]
Sorry for delayed answer from my side too. It was long time ago when I checked this code. I'm still interested to get into main branch and next release.

File: drivers/ethernet/eth_w5500.c (hunk 4 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:348]: It seems that when writing the comment marking the issue resolved then the comment itself is lost. This was my answer:
`generate_mac` is private field and used to assign random MAC. Random MAC that changes on every restart was something that I wanted to avoid in my application. In my solution I'm storing the MAC address in application and updating the driver MAC address on application startup.

**Comment by joelguittet** [comment]
I express my interest for the local-mac-address support for this driver. Will be very nice if can be merged for v3.3.0.

**Comment by giobauermeister** [comment]
I tested the changes in this pull request and it seems to be working.

**Review by teburd — DISMISSED** [review]

**Comment by teburd** [comment]
@jfischer-no

File: drivers/ethernet/eth_w5500.c [hunk 5]
```diff
@@ -433,14 +457,16 @@ static void w5500_set_macaddr(const struct device *dev)
...
-	ctx->mac_addr[1] = WIZNET_OUI_B1;
-	ctx->mac_addr[2] = WIZNET_OUI_B2;
 	if (ctx->generate_mac) {
 		ctx->generate_mac(ctx->mac_addr);
 	}
+#endif
+	/* override vendor bytes for both random and local_mac_address */
+	ctx->mac_addr[0] = WIZNET_OUI_B0;
+	ctx->mac_addr[1] = WIZNET_OUI_B1;
+	ctx->mac_addr[2] = WIZNET_OUI_B2;
```
Comment (jfischer-no) [inline drivers/ethernet/eth_w5500.c:469]: That looks weird and IMO this changes should be done in a separate commit.

`w5500_random_mac()` passes `WIZNET_OUI_B?` to `gen_random_mac()`, why is it necessary to overwrite it again?
Is it okay to assign only mac_addr[0...2] if there is no `DT_INST_PROP(0, zephyr_random_mac_address)` and why?

File: drivers/ethernet/eth_w5500.c [hunk 6]
```diff
@@ -342,28 +342,52 @@ static int w5500_set_config(const struct device *dev,
...
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
+		memcpy(ctx->mac_addr,
+			config->mac_address.addr,
+			sizeof(ctx->mac_addr));
+		/* override vendor bytes */
+		ctx->mac_addr[0] = WIZNET_OUI_B0;
+		ctx->mac_addr[1] = WIZNET_OUI_B1;
+		ctx->mac_addr[2] = WIZNET_OUI_B2;
```
Comment (jfischer-no) [inline drivers/ethernet/eth_w5500.c:355]: Why does it "override vendor byte"? Why can a user not set his own OUI?

File: drivers/ethernet/eth_w5500.c [hunk 7]
```diff
@@ -342,28 +342,52 @@ static int w5500_set_config(const struct device *dev,
...
+			dev->name,
+			ctx->mac_addr[0], ctx->mac_addr[1],
+			ctx->mac_addr[2], ctx->mac_addr[3],
+			ctx->mac_addr[4], ctx->mac_addr[5]);
+
+		/* Register Ethernet MAC Address with the upper layer */
+		net_if_set_link_addr(ctx->iface, ctx->mac_addr,
+			sizeof(ctx->mac_addr),
+			NET_LINK_ETHERNET);
+		return 0;
```
Comment (jfischer-no) [inline drivers/ethernet/eth_w5500.c:367]:
```c
			NET_LINK_ETHERNET);

		return 0;
```

File: drivers/ethernet/eth_w5500.c [hunk 8]
```diff
@@ -342,28 +342,52 @@ static int w5500_set_config(const struct device *dev,
...
+	default:
+		w5500_spi_read(dev, W5500_S0_MR, &mode, 1);
+		if (IS_ENABLED(CONFIG_NET_PROMISCUOUS_MODE) &&
+		    type == ETHERNET_CONFIG_TYPE_PROMISC_MODE) {
+			if (config->promisc_mode) {
+				if (!(mode & BIT(mr))) {
+					return -EALREADY;
+				}
+			}
```
Comment (jfischer-no) [inline drivers/ethernet/eth_w5500.c:376]: Stray changes?

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (joelguittet) [reply drivers/ethernet/eth_w5500.c:355]: @jfischer-no correct me if i'm wrong but I think this is also what is done for other Ethernet drivers in Zephyr?

Comment (jfischer-no) [reply drivers/ethernet/eth_w5500.c:355]: It seems to be a rumor. What driver does it for ETHERNET_CONFIG_TYPE_MAC_ADDRESS?

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:355]: If I remember correctly then I tried to set my own OUI bytes from my application too, but these values were ignored by the W5500 chip.

File: drivers/ethernet/eth_w5500.c (hunk 5 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:469]: I touched and tested this code long time ago. If I remember correctly then the reason to override here was that my W5500 hardware chip ignored custom OUI vendor bytes but underlying packet handling and filtering in Zephyr uses provided mac_addr. 

In example if user provided wrong OUI bytes in local_mac_address then the networking didn't work at all.

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:469]: But you are right, mac_addr[0...2] assignment could be moved to #else block if zephyr_random_mac_address is not used at the moment.

The reason I implemented it in that way was that I didn't fully understand why the original code was implemented in a way as it was. In example why to check if ctx->generate_mac is assigned when it is always there and assign the OUI bytes before it, when ctx->generate_mac() call assigns them anyway later. This all rendered the DT provided local_mac_address unusable and random MAC was used always.

The only explanation I could come up was that the original implementer wanted to make ctx->generate_mac function assignable from the application code or somewhere else and for this reason it is safer to override OUI bytes every time.

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (joelguittet) [reply drivers/ethernet/eth_w5500.c:355]: @jfischer-no as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_enc28j60.c#L820

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:355]: If W5500 hardware uses one MAC and the Zephyr stack knows another MAC in filtering layer then the networking doesn't work at all. I thought that it is better to fix it in driver level in the same way as the underlying HW is doing and saving the future developer using W5500 from trying to debug and understand why the network is not working.

I thought about another alternitive to return from this function with some error code if user provided OUI doesn't match with Wiznet OUI, but this means that Zephyr application has to be aware of underlying hardware and know about Wiznet OUI bytes which are not exposed from Zephyr headers.

File: drivers/ethernet/eth_w5500.c (hunk 8 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:376]: This was the original code I didn't want to touch or change in this PR as it wasn't fully clear to me why it was implemented in that way. This is what I wrote in another comment:
> I checked the code and promiscuous mode (MAC filter off) is activated by default in w5500_hw_start().
> 
> The bad thing is that whenever w5500_set_config() is called and the type is not ETHERNET_CONFIG_TYPE_PROMISC_MODE then promiscuous mode is disabled (MAC filter is turned on).
> 
> This needs more investigation and I didn't want to change this code with this PR.

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (joelguittet) [reply drivers/ethernet/eth_w5500.c:355]: I prefer overwriting as proposed in the PR in this case.

Comment (jfischer-no) [reply drivers/ethernet/eth_w5500.c:355]: > @jfischer-no as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_enc28j60.c#L820

No, it is not done during ETHERNET_CONFIG_TYPE_MAC_ADDRESS call.

> If W5500 hardware uses one MAC and the Zephyr stack knows another MAC in filtering layer then the networking doesn't work at all. I thought that it is better to fix it in driver level in the same way as the underlying HW is doing and saving the future developer using W5500 from trying to debug and understand why the network is not working.
> 
> I thought about another alternitive to return from this function with some error code if user provided OUI doesn't match with Wiznet OUI, but this means that Zephyr application has to be aware of underlying hardware and know about Wiznet OUI bytes which are not exposed from Zephyr headers.

If this controller can not change MAC address then this driver should not implement ETHERNET_CONFIG_TYPE_MAC_ADDRESS. If OUI is not modifiable, then the driver should not write to these registers/addresses. Overwriting OUI part multiple times in the driver is just confusing and opaque to maintainers and future contributors.

File: drivers/ethernet/eth_w5500.c (hunk 5 above)
Comment (jfischer-no) [reply drivers/ethernet/eth_w5500.c:469]: If there is no zephyr_random_mac_address property  and there is a valid MAC address stored in the controller registers, the driver should read MAC address from the controller registers.

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:355]: W5500 supports user assigned MAC. This is how random MAC works and how I'm using to send fixed MAC with ETHERNET_CONFIG_TYPE_MAC_ADDRESS from my application in production. W5500 ignores the OUI part in the MAC and sending whole MAC was implemented in the driver before I did my changes. I haven't checked from W5500 documentation by myself if it supports changing partial MAC. Somewhat I doubt that it supports it.

I agree that overwriting the OUI part in multiple places is not the best solution. But the original code was already confusing for me so I decided not to change or touch these parts as I wasn't sure why they were implemented in that was as they are in source code. This might need a separate PR to refactor the existing code which I didn't want to do in my PR.

What I could offer to do in my PR to improve the comments with the reasons why some parts of the code have done by my as they are, so that future maintainers and contributors could take this into account when refactoring the code.

File: drivers/ethernet/eth_w5500.c (hunk 5 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:469]: This is a good idea not to change the MAC if W5500 already has fixed and unique address stored internally. This was my original requirement to get unique and stable MAC and I somewhat remember that I even tried to test it out and it didn't work. I can't be sure at the moment as it all happened 7 months ago. But this should be a separate PR then as my PR is about making possible to change existing MAC from the user space application even if it was assigned in HW or set as random in driver.

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (jfischer-no) [reply drivers/ethernet/eth_w5500.c:355]: > W5500 supports user assigned MAC. This is how random MAC works and how I'm using to send fixed MAC with ETHERNET_CONFIG_TYPE_MAC_ADDRESS from my application in production. W5500 ignores the OUI part in the MAC and sending whole MAC was implemented in the driver before I did my changes. I haven't checked from W5500 documentation by myself if it supports changing partial MAC. Somewhat I doubt that it supports it.
>

"W5500 supports user assigned MAC" contradicts "W5500 ignores the OUI part in the MAC". Discussion here is not about random MAC but about implementation of ETHERNET_CONFIG_TYPE_MAC_ADDRESS setting. 

> I agree that overwriting the OUI part in multiple places is not the best solution. But the original code was already confusing for me so I decided not to change or touch these parts as I wasn't sure why they were implemented in the way as they are in source code. This might need a separate PR to refactor the existing code which I didn't want to do in my PR.
> 
> What I could offer to do in my PR is to improve the comments with the reasons why some parts of the code have done by my as they are, so that future maintainers and contributors could take this into account when refactoring the code.

You have not checked the manual but try to reuse some already dirty parts, it does not sound like the right approach.

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:355]: I have full respect for the original contributors and maintainers of this driver and thankful for the whole Zephyr project. I believe that there are reasons why every part of the code has evolved historically as it is in use today. Given the time constraints I had back then and have today this was the safest approach to reuse existing code from this driver and other drivers within Zephyr project (ETHERNET_CONFIG_TYPE_MAC_ADDRESS implementation for example) to achieve the goals for my project and not to touch or modify the existing code paths that I couldn't understand fully without having to dig into W5500 documentation details. I shared my solution in the hope that someone more knowledgeable in Zephyr and W5500 internals could find it useful or use the parts of my code to reimplement the missing functionality properly if needed.

File: drivers/ethernet/eth_w5500.c (hunk 7 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c:364]: Applied the proposal.

File: drivers/ethernet/eth_w5500.c (hunk 5 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: I tested the custom OUI bytes again and this time there were no issues. No idea, if I did something wrong during testing 7 months ago or had different W5500 HW revision back then when I did the first tests. Reverted my changes to original code but kept the random mac address DT conditional define.

File: drivers/ethernet/eth_w5500.c (hunk 6 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: Tested again and it looks like that this time custom OUI bytes were accepted with my W5500 chip. Removed the override OUI bytes from this PR.

Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: To clarify one statement I made yesterday, I checked the W5500 datasheet and I was wrong and W5500 should support partial MAC update as all settings are stored in one common register block.

File: drivers/ethernet/eth_w5500.c (hunk 5 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: > If there is no zephyr_random_mac_address property and there is a valid MAC address stored in the controller registers, the driver should read MAC address from the controller registers.

Checked this from datasheet and after reset the Source Hardware Address (MAC address) is filled with default values 0x00. This means that MAC address has to be updated after every restart from the driver or application.

File: drivers/ethernet/eth_w5500.c [hunk 9]
```diff
@@ -342,28 +342,49 @@ static int w5500_set_config(const struct device *dev,
...
-	} else {
-		if (mode & BIT(mr)) {
-			return -EALREADY;
+			/* set */
+			WRITE_BIT(mode, mr, 1);
 		}
 
-		/* set */
-		WRITE_BIT(mode, mr, 1);
+		return w5500_spi_write(dev, W5500_S0_MR, &mode, 1);
```
[the remark this answers is not in the document]
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: @jfischer-no took your changes and added some comments. 

Also changed the w5500_hw_start() so that W5500 is started by default with promiscuous mode disabled (hw MAC filtering enabled). So far hw promiscuous mode was active from start even if CONFIG_NET_PROMISCUOUS_MODE was disabled. Don't know if this could pose any security risks in the past.

Don't have any tools to test promiscuous mode separately and this part of the code wasn't tested.

File: drivers/ethernet/eth_w5500.c (hunk 8 above)
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: Disabled promiscuous mode (MAC filter disabled) in w5500_hw_start().

File: drivers/ethernet/eth_w5500.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (rpaat) [reply drivers/ethernet/eth_w5500.c]: Moved promiscuous mode changes to separate commit.

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
